File: backend/app/schemas/local_video.py
"""Local Video API schemas (FEATURE-CREATORS-V2 / C2-S1)."""

from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, Field, field_validator

from app.core.local_video_constants import (
    ALLOWED_LOCAL_VIDEO_CONTENT_TYPES,
    LOCAL_VIDEO_COMMENT_BODY_MAX_LENGTH,
    LOCAL_VIDEO_DEFAULT_CITY,
    LOCAL_VIDEO_DESCRIPTION_MAX_LENGTH,
    LOCAL_VIDEO_MAX_BYTES,
    LOCAL_VIDEO_TITLE_MAX_LENGTH,
    LocalVideoReportReason,
    LocalVideoStatus,
    LocalVideoType,
)


class LocalVideoUploadInitRequest(BaseModel):
    filename: str = Field(min_length=1, max_length=255)
    content_type: str
    file_size_bytes: int = Field(gt=0)

    @field_validator("content_type")
    @classmethod
    def validate_content_type(cls, value: str) -> str:
        normalized = value.strip().lower()
        if normalized not in ALLOWED_LOCAL_VIDEO_CONTENT_TYPES:
            raise ValueError("Type de fichier vidéo non supporté.")
        return normalized

    @field_validator("file_size_bytes")
    @classmethod
    def validate_file_size(cls, value: int) -> int:
        if value > LOCAL_VIDEO_MAX_BYTES:
            raise ValueError("Fichier trop volumineux.")
        return value


class LocalVideoUploadInitResponse(BaseModel):
    upload_id: uuid.UUID
    presigned_url: str
    storage_key: str
    expires_at: datetime
    upload_method: str
    upload_headers: dict[str, str]


class LocalVideoPublishRequest(BaseModel):
    upload_id: uuid.UUID
    city: str = Field(default=LOCAL_VIDEO_DEFAULT_CITY, min_length=1, max_length=64)
    neighborhood_id: uuid.UUID
    video_type: LocalVideoType
    title: str | None = Field(default=None, max_length=LOCAL_VIDEO_TITLE_MAX_LENGTH)
    description: str | None = Field(
        default=None,
        max_length=LOCAL_VIDEO_DESCRIPTION_MAX_LENGTH,
    )
    cultural_place_id: uuid.UUID | None = None
    local_event_id: uuid.UUID | None = None
    tribe_id: uuid.UUID | None = None
    organization_id: uuid.UUID | None = None
    latitude: float | None = Field(default=None, ge=-90, le=90)
    longitude: float | None = Field(default=None, ge=-180, le=180)


class LocalVideoItem(BaseModel):
    id: uuid.UUID
    author_user_id: uuid.UUID
    city: str
    neighborhood_id: uuid.UUID
    video_type: LocalVideoType
    title: str | None
    description: str | None
    cultural_place_id: uuid.UUID | None
    local_event_id: uuid.UUID | None
    tribe_id: uuid.UUID | None
    organization_id: uuid.UUID | None
    media_url: str
    thumbnail_url: str
    duration_seconds: float
    file_size_bytes: int
    mime_type: str
    latitude: float | None
    longitude: float | None
    status: LocalVideoStatus
    processing_error: str | None
    published_at: datetime | None
    created_at: datetime


class LocalVideoFeedAuthor(BaseModel):
    id: uuid.UUID
    username: str | None
    full_name: str
    avatar_url: str | None


class LocalVideoFeedItem(BaseModel):
    id: uuid.UUID
    author_user_id: uuid.UUID
    author: LocalVideoFeedAuthor
    city: str
    neighborhood_id: uuid.UUID
    neighborhood_name: str
    neighborhood_slug: str
    video_type: LocalVideoType
    title: str | None
    description: str | None
    cultural_place_id: uuid.UUID | None
    cultural_place_slug: str | None
    cultural_place_name: str | None
    local_event_id: uuid.UUID | None
    tribe_id: uuid.UUID | None
    organization_id: uuid.UUID | None
    media_url: str
    thumbnail_url: str
    duration_seconds: float
    mime_type: str
    latitude: float | None
    longitude: float | None
    status: LocalVideoStatus
    published_at: datetime | None
    created_at: datetime
    distance_meters: int | None
    walk_minutes: int | None
    like_count: int
    comment_count: int
    liked_by_me: bool = False


class LocalVideoFeedResponse(BaseModel):
    items: list[LocalVideoFeedItem]
    next_cursor: str | None
    city: str


class LocalVideoLikeResponse(BaseModel):
    liked: bool
    like_count: int


class LocalVideoCommentCreateRequest(BaseModel):
    body: str = Field(..., min_length=1, max_length=LOCAL_VIDEO_COMMENT_BODY_MAX_LENGTH)

    @field_validator("body")
    @classmethod
    def strip_body(cls, value: str) -> str:
        return value.strip()


class LocalVideoCommentResponse(BaseModel):
    id: uuid.UUID
    video_id: uuid.UUID
    author_user_id: uuid.UUID
    author_display_name: str
    author_username: str | None
    body: str
    created_at: datetime
    updated_at: datetime


class LocalVideoCommentListResponse(BaseModel):
    items: list[LocalVideoCommentResponse]
    next_cursor: str | None = None


class LocalVideoReportCreateRequest(BaseModel):
    reason: LocalVideoReportReason

File: frontend/packages/types/src/local-video.ts
/** Local Video V2 — types API (FEATURE-CREATORS-V2 / VIDEO-04A). */

export type LocalVideoTypeId =
  | "bon_plan"
  | "moment"
  | "quartier"
  | "lieu"
  | "tribu"
  | "autre";

export type LocalVideoStatusId =
  | "processing"
  | "published"
  | "failed"
  | "hidden"
  | "deleted";

/** Pipeline worker (VIDEO-03A) — aligné backend `LocalVideoProcessingStatus`. */
export type LocalVideoProcessingStatusId =
  | "uploaded"
  | "processing"
  | "ready"
  | "failed";

export type LocalVideoContentType = "video/mp4" | "video/quicktime";

export const LOCAL_VIDEO_MAX_BYTES = 52_428_800;
export const LOCAL_VIDEO_MAX_DURATION_SECONDS = 90;
export const LOCAL_VIDEO_ALLOWED_CONTENT_TYPES: readonly LocalVideoContentType[] = [
  "video/mp4",
  "video/quicktime",
] as const;

export type LocalVideoAuthor = {
  id: string;
  username: string | null;
  full_name: string;
  avatar_url: string | null;
};

/** Détail vidéo — `GET /local-videos/{id}` (`LocalVideoItem` backend). */
export type LocalVideo = {
  id: string;
  author_user_id: string;
  city: string;
  neighborhood_id: string;
  video_type: LocalVideoTypeId;
  title: string | null;
  description: string | null;
  cultural_place_id: string | null;
  local_event_id: string | null;
  tribe_id: string | null;
  organization_id: string | null;
  media_url: string;
  thumbnail_url: string;
  duration_seconds: number;
  file_size_bytes: number;
  mime_type: string;
  latitude: number | null;
  longitude: number | null;
  status: LocalVideoStatusId;
  processing_status: LocalVideoProcessingStatusId;
  processing_error: string | null;
  published_at: string | null;
  created_at: string;
};

/** Session upload — `POST /local-videos/upload-init` (`LocalVideoUploadInitResponse`). */
export type LocalVideoUpload = {
  upload_id: string;
  presigned_url: string;
  storage_key: string;
  expires_at: string;
  upload_method: string;
  upload_headers: Record<string, string>;
};

export type LocalVideoUploadInitPayload = {
  filename: string;
  content_type: LocalVideoContentType;
  file_size_bytes: number;
  city?: string;
  neighborhood_id?: string;
  organization_id?: string;
};

export type LocalVideoPublishPayload = {
  upload_id: string;
  city?: string;
  neighborhood_id: string;
  video_type: LocalVideoTypeId;
  title?: string | null;
  description?: string | null;
  cultural_place_id?: string | null;
  local_event_id?: string | null;
  tribe_id?: string | null;
  organization_id?: string | null;
  latitude?: number | null;
  longitude?: number | null;
};

/** Réponse publish — HTTP 202 (`LocalVideoPublishAcceptedResponse`). */
export type LocalVideoPublishAcceptedResponse = {
  id: string;
  status: LocalVideoStatusId;
  processing_status: LocalVideoProcessingStatusId;
  job_id: string;
  message: string;
};

export type LocalVideoFeedItem = {
  id: string;
  author_user_id: string;
  author: LocalVideoAuthor;
  city: string;
  neighborhood_id: string;
  neighborhood_name: string;
  neighborhood_slug: string;
  video_type: LocalVideoTypeId;
  title: string | null;
  description: string | null;
  cultural_place_id: string | null;
  cultural_place_slug: string | null;
  cultural_place_name: string | null;
  local_event_id: string | null;
  tribe_id: string | null;
  organization_id: string | null;
  media_url: string;
  thumbnail_url: string;
  duration_seconds: number;
  mime_type: string;
  latitude: number | null;
  longitude: number | null;
  status: LocalVideoStatusId;
  published_at: string | null;
  created_at: string;
  distance_meters: number | null;
  walk_minutes: number | null;
  like_count: number;
  comment_count: number;
  view_count: number;
  liked_by_me: boolean;
};

export type LocalVideoLikeResponse = {
  liked: boolean;
  like_count: number;
};

export type LocalVideoComment = {
  id: string;
  video_id: string;
  author_user_id: string;
  author_display_name: string;
  author_username: string | null;
  body: string;
  created_at: string;
  updated_at: string;
};

export type LocalVideoCommentListResponse = {
  items: LocalVideoComment[];
  next_cursor: string | null;
};

export type LocalVideoCommentCreatePayload = {
  body: string;
};

export type LocalVideoReportReason =
  | "spam"
  | "harassment"
  | "hate"
  | "violence"
  | "sexual"
  | "copyright"
  | "other";

export type LocalVideoReportCreatePayload = {
  reason: LocalVideoReportReason;
};

export type LocalVideoListParams = {
  cursor?: string;
  limit?: number;
  city?: string;
  latitude?: number;
  longitude?: number;
};

export type LocalVideoListResponse = {
  items: LocalVideoFeedItem[];
  next_cursor: string | null;
  city: string;
};

/** Codes d'erreur API Local Video connus (backend `AppError.code`). */
export type LocalVideoErrorCode =
  | "LOCAL_VIDEO_INVALID_CONTENT"
  | "LOCAL_VIDEO_INVALID_TYPE"
  | "LOCAL_VIDEO_TOO_LARGE"
  | "LOCAL_VIDEO_SIZE_MISMATCH"
  | "LOCAL_VIDEO_UPLOAD_EXPIRED"
  | "LOCAL_VIDEO_UPLOAD_MISSING"
  | "LOCAL_VIDEO_UPLOAD_NOT_AVAILABLE"
  | "LOCAL_VIDEO_UPLOAD_ALREADY_USED"
  | "LOCAL_VIDEO_UPLOAD_NOT_FOUND"
  | "LOCAL_VIDEO_FORBIDDEN"
  | "LOCAL_VIDEO_NOT_FOUND"
  | "LOCAL_VIDEO_INVALID_NEIGHBORHOOD"
  | "LOCAL_VIDEO_CITY_MISMATCH"
  | "LOCAL_VIDEO_CITY_SLUG_MISMATCH"
  | "LOCAL_VIDEO_INVALID_MEDIA"
  | "LOCAL_VIDEO_TOO_LONG"
  | "LOCAL_VIDEO_TRANSCODE_FAILED"
  | "LOCAL_VIDEO_THUMBNAIL_FAILED"
  | "LOCAL_VIDEO_BINARY_ENDPOINT_UNAVAILABLE"
  | "LOCAL_VIDEO_PROCESSING_TIMEOUT"
  | "RATE_LIMITED"
  | "UNKNOWN_ERROR";
